federation/pkg/federation-controller/service/dns.go

```diff
@@ -16,25 +16,230 @@ limitations under the License.
+	return "zone-of-cluster-" + clusterName, nil
 }
 
 // getClusterRegionName returns the name of the region where the specified cluster exists (e.g. us-east1 on GCE, or "us-east-1" on AWS)
-func getClusterRegionName(clusterName string) string {
+func (s *ServiceController) getClusterRegionName(clusterName string) (string, error) {
 	// TODO: quinton: Get this from the cluster API object - from the annotation on a node in the cluster - it doesn't contain this yet.
-	return "region-of-cluster-" + clusterName
+	return "region-of-cluster-" + clusterName, nil
 }
 
 // getFederationDNSZoneName returns the name of the managed DNS Zone configured for this federation
-func getFederationDNSZoneName() string {
-	return "mydomain.com" // TODO: quinton: Get this from the federation configuration.
+func (s *ServiceController) getFederationDNSZoneName() (string, error) {
+	return "example.com", nil // TODO: quinton: Get this from the federation configuration.
+	// Note: For unit testing this must match the domain populated in the test/stub dnsprovider.
+}
+
+// getZone is a hack around the fact that dnsprovider does not yet support a Get() method, only a List() method.  TODO:  Fix that.
+func getZone(dnsZoneName string, zonesInterface dnsprovider.Zones) (dnsprovider.Zone, error) {
+	zones, err := zonesInterface.List()
+	if err != nil {
+		return nil, err
+	}
+	for _, zone := range zones {
+		if zone.Name() == dnsZoneName {
+			return zone, nil
+		}
+	}
+	return nil, fmt.Errorf("DNS zone %s not found.", dnsZoneName)
 }
 
-func ensureDNSRecords(clusterName string, cachedService *cachedService) error {
+// getRrset is a hack around the fact that dnsprovider.ResourceRecordSets interface does not yet include a Get() method, only a List() method.  TODO:  Fix that.
+func getRrset(dnsName string, rrsetsInterface dnsprovider.ResourceRecordSets) (dnsprovider.ResourceRecordSet, error) {
+	rrsets, err := rrsetsInterface.List()
+	if err != nil {
+		return nil, err
+	}
+	for _, rrset := range rrsets {
+		if rrset.Name() == dnsName {
+			return rrset, nil
+		}
+	}
+	return nil, fmt.Errorf("Rrset %s not found.", dnsName)
+}
+
+/* ensureDnsRrsets ensures (idempotently, and with minimum mutations) that all of the DNS resourcerecordsets for dnsName are consistent with endpoints.
+   if endpoints is nil or empty, a CNAME record to uplevelCname is ensured
+*/
+func (s *ServiceController) ensureDnsRrsets(dnsZoneName, dnsName string, endpoints []string, uplevelCname string) error {
+	zone, err := getZone(dnsZoneName, s.zones) // TODO: s.zones.Get(dnsZoneName)
+	if err != nil {
+		return err
+	}
+	rrsets, supported := zone.ResourceRecordSets()
+	if !supported {
+		return fmt.Errorf("Failed to ensure DNS records for %s. DNS provider does not support the ResourceRecordSets interface.", dnsName)
+	}
+	rrset, err := getRrset(dnsName, rrsets) // TODO: rrsets.Get(dnsName) // TODO: quinton - need more params here.
+	if err != nil {
+		// Assume that it doesn't exist yet, so create it
```

Comment:
This should not be assumed but should actually be verified. What if getRrset is failing for auth reason or something that cannot be fixed by creating records, but instead attempting to do so might aggravate the problem? The right approach here is to define `NotFoundErr` constants of type `error` in this package and return them in your `getRrset()` and `getZone()`